### PR TITLE
[ntuple] Add S3 anchor and its tests for S3 backend

### PR DESCRIPTION
**Pull request by JasMehta08**
# This Pull request:
(Is a part of the GSoC 2026 project `S3 Backend for RNTuple.`)

Creates the anchor for the S3 backend.
## Changes or fixes:

- Adds `S3NTupleAnchor` struct, using JSON to store the anchor data. 
- Adds `ROOT_BUILD_OPTION(s3 OFF "Enable RNTuple support for S3-compatible object storage")`to conditionally compile.
- Adds tests that cover round-trip, version guard, malformed input, type mismatches, boundary values, and forward compatibility.

## Checklist:

- [x] tested changes locally

### CODE REVIEW

**Comment by JasMehta08** [comment]
@jblomer 

I have made the changes so that the build for S3 backend only occurs when curl is present, and removed the s3 flag.

File: tree/ntuple/inc/ROOT/RPageStorageS3.hxx [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+\ingroup NTuple
+\brief Entry point for an RNTuple stored in S3-compatible object storage.
+
+The anchor is serialized as a JSON object and stored at the base URL of the ntuple.
+It contains the information needed to locate and read the header and footer envelopes.
+The anchor is always the last object written during CommitDatasetImpl, ensuring atomicity:
+if the anchor exists, the entire ntuple is complete.
+*/
+// clang-format on
+struct RS3NTupleAnchor {
```
Comment (jblomer) [inline tree/ntuple/inc/ROOT/RPageStorageS3.hxx]: Maybe better:
```suggestion
struct RNTupleAnchorS3 {
```

File: tree/ntuple/inc/ROOT/RPageStorageS3.hxx [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+   std::uint16_t fVersionPatch = RNTuple::kVersionPatch;
+   /// Pattern for resolving object IDs to full S3 URLs.
+   /// ${baseurl} is replaced with the anchor URL, ${objid} with the numeric object ID.
+   std::string fUrlTemplate;
+   /// Object ID and byte offset of the compressed header within the S3 object
+   std::uint64_t fHeaderObjId = 0;
+   std::uint64_t fHeaderOffset = 0;
+   /// Compressed and uncompressed sizes of the header envelope
+   std::uint32_t fNBytesHeader = 0;
+   std::uint32_t fLenHeader = 0;
```
Comment (jblomer) [inline tree/ntuple/inc/ROOT/RPageStorageS3.hxx]: Here and for the footer: since we use 64bits in the regular anchor, let's do the same here.

File: tree/ntuple/inc/ROOT/RPageStorageS3.hxx [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+   std::uint64_t fFooterObjId = 0;
+   std::uint64_t fFooterOffset = 0;
+   /// Compressed and uncompressed sizes of the footer envelope
+   std::uint32_t fNBytesFooter = 0;
+   std::uint32_t fLenFooter = 0;
+
+   bool operator==(const RS3NTupleAnchor &other) const;
+
+   /// Serialize the anchor to a JSON string suitable for storage at the base URL
+   std::string Serialize() const;
```
Comment (jblomer) [inline tree/ntuple/inc/ROOT/RPageStorageS3.hxx]:
```suggestion
   std::string ToJSON() const;
```

File: tree/ntuple/inc/ROOT/RPageStorageS3.hxx [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+   /// Deserialize the anchor from a JSON string. Returns an error on malformed or incompatible input.
+   RResult<void> Deserialize(const std::string &json);
```
Comment (jblomer) [inline tree/ntuple/inc/ROOT/RPageStorageS3.hxx]:
```suggestion
   RResult<void> FromJSON(const std::string &json);
```

File: tree/ntuple/src/RPageStorageS3.cxx [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+   jsonAnchor["format_version_patch"] = fVersionPatch;
+   jsonAnchor["url_template"] = fUrlTemplate;
+   jsonAnchor["header_obj_id"] = fHeaderObjId;
+   jsonAnchor["header_offset"] = fHeaderOffset;
+   jsonAnchor["nbytes_header"] = fNBytesHeader;
+   jsonAnchor["len_header"] = fLenHeader;
+   jsonAnchor["footer_obj_id"] = fFooterObjId;
+   jsonAnchor["footer_offset"] = fFooterOffset;
+   jsonAnchor["nbytes_footer"] = fNBytesFooter;
+   jsonAnchor["len_footer"] = fLenFooter;
```
Comment (jblomer) [inline tree/ntuple/src/RPageStorageS3.cxx]: Maybe, since we use camelCase in the C++ code, let's do the same in JSON.

File: tree/ntuple/CMakeLists.txt [hunk 6]
```diff
@@ -121,6 +121,14 @@ if(daos OR daos_mock)
   endif()
 endif()
 
+# Enable RNTuple support for S3-compatible object storage
+if(curl)
+  set(ROOTNTuple_EXTRA_HEADERS ${ROOTNTuple_EXTRA_HEADERS} ROOT/RPageStorageS3.hxx)
+  target_sources(ROOTNTuple PRIVATE src/RPageStorageS3.cxx)
+  target_compile_definitions(ROOTNTuple PRIVATE R__ENABLE_S3)
```
Comment (jblomer) [inline tree/ntuple/CMakeLists.txt]: Let's wait until we actually need the define.

File: tree/ntuple/inc/ROOT/RPageStorageS3.hxx (hunk 4 above)
Comment (jblomer) [inline tree/ntuple/inc/ROOT/RPageStorageS3.hxx]: I think it is easier to make this method static and return a `RResult<RS3NTupleAnchor>`. Then we don't need to care about object reset.

File: tree/ntuple/inc/ROOT/RPageStorageS3.hxx [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+The anchor is serialized as a JSON object and stored at the base URL of the ntuple.
+It contains the information needed to locate and read the header and footer envelopes.
+The anchor is always the last object written during CommitDatasetImpl, ensuring atomicity:
+if the anchor exists, the entire ntuple is complete.
+*/
+// clang-format on
+struct RS3NTupleAnchor {
+   /// Allows evolving the anchor JSON schema in future versions
+   std::uint32_t fVersionAnchor = 1;
```
Comment (jblomer) [inline tree/ntuple/inc/ROOT/RPageStorageS3.hxx]: Not sure if we gain anything by reserving version 0. I think we can just start with it.

**Review by jblomer — COMMENTED** [review]
Very nice!

**Comment by JasMehta08** [comment]
@jblomer 

I have made the update according to the comments

File: tree/ntuple/inc/ROOT/RPageStorageS3.hxx [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+   /// Compressed and uncompressed sizes of the footer envelope
+   std::uint64_t fNBytesFooter = 0;
+   std::uint64_t fLenFooter = 0;
+
+   bool operator==(const RNTupleAnchorS3 &other) const;
+
+   /// Serialize the anchor to a JSON string suitable for storage at the base URL
+   std::string ToJSON() const;
+   /// Deserialize the anchor from a JSON string. Returns an error on malformed or incompatible input.
+   static RResult<RNTupleAnchorS3> FromJSON(const std::string &json);
```
Comment (jblomer) [inline tree/ntuple/inc/ROOT/RPageStorageS3.hxx]: One more tiny comment: perhaps better
```suggestion
   static RResult<RNTupleAnchorS3> CreateFromJSON(const std::string &json);
```
To make it clear that a new object comes out of this function.

**Review by jblomer — APPROVED** [review]
LGTM (but see the clang-format report)

**Comment by JasMehta08** [comment]
@jblomer

I have made the change in the small comment which had issue with Clang format

**Comment by JasMehta08** [comment]
@jblomer 

Please have a review and let me know if there are any changes remaining, 

Thanks!
